File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>>ans;
        stack<TreeNode*>s1;
        stack<TreeNode*>s2;
        if(root==NULL)return ans;
        s1.push(root);
        while(!s1.empty() || !s2.empty()){
            vector<int>v;
            if(!s1.empty()){
                while(!s1.empty()){
                    TreeNode* temp=s1.top();s1.pop();
                    if(temp->left)s2.push(temp->left);
                    if(temp->right)s2.push(temp->right);
                    v.push_back(temp->val);
                }
            }else{
                while(!s2.empty()){
                    TreeNode* temp=s2.top();s2.pop();
                    if(temp->right)s1.push(temp->right);
                    if(temp->left)s1.push(temp->left);
                    v.push_back(temp->val);
                }

            }
            ans.push_back(v);
        }
        return ans;
        
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        List<List<Integer>>result = new ArrayList<>();
        if(root==null){
            return result;
        }
        q.add(root);
        int k=0;
        while(!q.isEmpty()){
            int size=q.size();
            List<Integer>level = new ArrayList<>();
            for(int i=0;i<size;i++){
                TreeNode curr = q.poll();
                level.add(curr.val);
                if(curr.left!=null){
                    q.add(curr.left);
                }
                if(curr.right!=null){
                    q.add(curr.right);
                }
            }
            if(k%2!=0){
                Collections.reverse(level);
            }
            result.add(new ArrayList<>(level));
            k++;
        }
        return result;
    }
}